### PR TITLE
Support more classify metrics

### PR DESCRIPTION
**Pull request by ScottMcMichael**

### CODE REVIEW

File: delta/extensions/losses.py [hunk 1]
```diff
@@ -28,11 +28,39 @@
...
+
 def ms_ssim(y_true, y_pred):
     """
-    `tf.image.ssim_multiscale` as a loss function.
-    """
-    return 1.0 - tf.image.ssim_multiscale(y_true, y_pred, 4.0)
+    `tf.image.ssim_multiscale` as a loss function. This loss function requires two
+    dimensional inputs.
+    """
+    y_true = tf.cond(tf.math.less(tf.rank(y_true), 3),
```
Comment (bcoltin) [inline delta/extensions/losses.py:52]: What is this for?

File: delta/extensions/metrics.py [hunk 2]
```diff
@@ -71,7 +71,6 @@ def __init__(self, label=None, class_id: int=None, name: str=None, binary: int=F
         self._total_class = self.add_weight('total_class', initializer='zeros')
         self._true_positives = self.add_weight('true_positives', initializer='zeros')
 
-    # sample_weight is unused but required by tensorflow
```
Comment (bcoltin) [inline delta/extensions/metrics.py:74]: can you put back this comment? Added this because we recently deleted sample_weight and broke things :)

File: delta/ml/predict.py [hunk 3]
```diff
@@ -471,7 +471,7 @@ def _process_block(self, pred_image, y, x, labels, label_nodata):
 
             if self._metrics:
                 for m in self._metrics:
-                    m.update_state(valid_labels.compressed(), valid_pred.compressed()) #pylint: disable=E1101
```
Comment (bcoltin) [inline delta/ml/predict.py:474]: I think you need the compressed to not count masked values... Is there a reason it was removed?

File: delta/subcommands/classify.py [hunk 4]
```diff
@@ -121,6 +121,10 @@ def __init__(self, loss_object):
 
     def update_state(self, y_true, y_pred, sample_weight=None): #pylint: disable=unused-argument, arguments-differ
         this_loss = self._loss_object.call(y_true, y_pred)
+        # In some situations the loss will not be a single number, take the mean to make it one
+        this_loss = tf.cond(tf.math.equal(tf.rank(this_loss), 0),
```
Comment (bcoltin) [inline delta/subcommands/classify.py:125]: The loss should not always be reduced to a single number. It is valid to have multiple dimensions. Sometimes it is important to preserve that even when reducing, if there are different batch sizes they will be waited differently in the reducer.

File: tests/test_extensions.py [hunk 5]
```diff
@@ -46,9 +46,11 @@ def test_gaussian_sample():
 
 def test_ms_ssim():
     l = ext.loss('ms_ssim')
-    assert l(tf.zeros((1, 180, 180, 1)), tf.zeros((1, 180, 180, 1))) == 0.0
+    assert tf.math.reduce_mean(l(tf.zeros((1, 180, 180, 1)),
```
Comment (bcoltin) [inline tests/test_extensions.py:49]: why?

**Review by bcoltin — CHANGES_REQUESTED** [review]

File: delta/extensions/losses.py (hunk 1 above)
Comment (ScottMcMichael) [reply delta/extensions/losses.py:52]: If the input is [x, y] it needs to be expanded to [x, y, 1].

File: delta/ml/predict.py (hunk 3 above)
Comment (ScottMcMichael) [reply delta/ml/predict.py:474]: The compressed call converts the data to a 1D array, making it incompatible with the msssim metrics.  I don't know if there is a good way to use msssim with masked values.

File: delta/subcommands/classify.py (hunk 4 above)
Comment (ScottMcMichael) [reply delta/subcommands/classify.py:125]: Do our metric classes in metrics.py also need to change?  They use reduce_add in a similar manner.

File: delta/ml/predict.py (hunk 3 above)
Comment (bcoltin) [reply delta/ml/predict.py:474]: Oh, good point, I hadn't thought of that! MSSSIM doesn't make much sense to use here... the more common case would be something like precision / recall, where we would definitely want to filter out the nodata pixels. I think it might be better to state we only support metrics on individual pixels here. What do you think @mvonpohle ?

File: delta/subcommands/classify.py (hunk 4 above)
Comment (bcoltin) [reply delta/subcommands/classify.py:125]: I had misunderstood what I was looking at, sorry! This is the right thing to do!

File: tests/test_extensions.py (hunk 5 above)
Comment (ScottMcMichael) [reply tests/test_extensions.py]: This turned out to be caused by a weird tensorflow behavior, so the test change has been reverted.
